### ISSUE TITLE
CUDA memory error when using python_backend with async 

### ISSUE THREAD

**Issue by seoungbae-park** [triton-inference-server/server#6070]
**Description**
My python_backend execute code is below 
![image](https://github.com/triton-inference-server/server/assets/94895901/3ab9b645-2331-4c6b-a376-06d5692befa3)
when i run without async doesn't have any issues,
but when i use async to run, cuda memory issue generates and triton shut downs.
 

![image](https://github.com/triton-inference-server/server/assets/94895901/ea3e2c9f-724b-4ca2-a402-72d6fe2637c3)
By the way, out of memory is too high 274GB? one model only takes 1~2GB and there is only two models inside "for loop"

**Triton Information**
What version of Triton are you using? 23.05

Are you using the Triton container or did you build it yourself? container

**To Reproduce**
Steps to reproduce the behavior.

Describe the models (framework, inputs, outputs), ideally include the model configuration file (if using an ensemble include the model configuration file for that as well).

**Expected behavior**
No memory issue. Only uses 3~5GB is enough for 2model concurrently

**Comment by jturner116** [comment]
https://github.com/triton-inference-server/server/issues/6051
Same error here I think

**Comment by Tabrizian** [comment]
From the logs it looks like the model/backend is trying to allocate 256GBs of memory. If you only have 274 GBs of memory available, I think it makes sense that you're not able to execute more than on request.

**Comment by seoungbae-park** [comment]
> From the logs it looks like the model/backend is trying to allocate 256GBs of memory. If you only have 274 GBs of memory available, I think it makes sense that you're not able to execute more than on request.

I mentioned that my model takes only 1~2GBs its no reason to allocate 256GBs of memory. I'm loading only 2~3models and each of model takes only 1~2GBs.

**Comment by Tabrizian** [comment]
@seoungbae-park If you run the inference outside Triton in a loop how much memory would it consume? It could be that the model itself is small but the outputs it generates are large causing this issue.

**Comment by seoungbae-park** [comment]
@Tabrizian No, it should not consume that much memory. This is my full code for inference and It does not inference outside triton. 
```
def inference_boxinst(self, image, pb_image, model_name, pre_nms_thresh, mask_threshold, request_id):
        start = time.time()
        pb_pre_nms_thresh = pb_utils.Tensor("pre_nms_thresh", pre_nms_thresh)
        pb_mask_threshold = pb_utils.Tensor("mask_threshold", mask_threshold)
        thsc_infer_request = pb_utils.InferenceRequest(
            model_name=model_name,
            requested_output_names=['boxes', 'scores', 'classes', 'masks', 'img_ids'],
            inputs=[pb_image, pb_pre_nms_thresh, pb_mask_threshold],
            request_id=request_id)

        thsc_infer_response = thsc_infer_request.exec()
        if thsc_infer_response.has_error():
            raise pb_utils.TritonModelException(thsc_infer_response.error().message())
        self.logger.log(f"[AI] {model_name} - ran in {round(time.time() - start, 3)}", self.logger.INFO)
        result = self.boxinst_output_to_np(image, model_name, thsc_infer_response)
        return result
```
